### PR TITLE
Multi thread proof of concept (Do not merge)

### PR DESCRIPTION
**Pull request by huwpascoe**
Has the dependency boost::lockfree which is not part of the externals.

This shows that the GPU core is able to run separately from the CPU core with little modification. I haven't made any attempt to optimize or understand the hardware, but hopefully this can give some inspiration.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/2955)

### CODE REVIEW

**Comment by wwylele** [comment]
I think the main concern about separate GPU into its own thread is that it can introduce indeterminacy, as it does in this easy implementation: the cycle of GPU thread to complete one task is not strongly tied to `CoreTiming` anymore, and depends on different hardware performance, the same game may behave different on different computes.

I don't have solid opinion to this yet, so I am just throwing here some questions with which we can discuss:

 - do we want to keep determinism? 
 - do we want accuracy? i.e. emulate the actual latency of a real 3DS GPU. Is it even possible to accurately measure the latency?
 - is undetermined latency considered more accurate than no latency?
 - can we have an option to switch between single thread/multithread?

**Comment by lioncash** [comment]
> the same game may behave different on different computes.

If this is the case, then the implementation should likely be improved to mitigate that, or be avoided for the time being. I can see this being kind of annoying to build a FifoCI analog around, if one is desirable (it should be), since consistent behavior is a necessity for accurate comparisons between changes. 

> do we want to keep determinism?

If we plan to support TASing in the future (or if someone else would like to add that feature), we'd likely want determinism.

**Comment by huwpascoe** [comment]
>do we want to keep determinism?
The VBlank Interrupt is probably important in this case. But are there other situations where the CPU has to wait for the GPU to send data back? If the model is entirely producer-consumer then the CPU thread decides what's what.

>can we have an option to switch between single thread/multithread?
That would be easy enough to do.

**Comment by wwylele** [comment]
> But are there other situations where the CPU has to wait for the GPU to send data back?

Yes, CPU can read back what GPU has rendered. For example, as I observed it is likely that in Monster Hunter games fetch back the rendered framebuffer and do an overall color balance on CPU side before displaying. Pokemon games might has CPU and GPU working together to draw the outlines as well.

**Comment by BreadFish64** [comment]
after testing Pokemon sun
At best, this implementation is (rarely) 10%ish faster.
Usually it's slower by up to 40-50%

**Comment by emmauss** [comment]
@BreadFish64  this isnt a PR for test. it is a proof of concept, showing that it is possible, but there havve not been anything implemented that takes full advantage of multithreading.

**Comment by BreadFish64** [comment]
@emmauss I'm aware, I just tried it for fun and @wwylele told me to drop the results here

File: src/core/core.cpp [hunk 1]
```diff
@@ -55,13 +70,20 @@ System::ResultStatus System::RunLoop(int tight_loop) {
         CoreTiming::Advance();
         PrepareReschedule();
     } else {
-        cpu_core->Run(tight_loop);
+        cpu_core->Run(TightLoop);
```
Comment (MerryMage) [inline src/core/core.cpp]: Race condition: `TightLoop` may be advanced between this statement and the next. Setting `TightLoop = 0;` thus seems like incorrect behaviour.

File: src/core/hw/gpu.cpp [hunk 2]
```diff
@@ -28,6 +29,14 @@ namespace GPU {
 
 Regs g_regs;
 
+struct WriteThing {
+    u32 addr;
+    u32 data;
+};
+
+std::atomic<bool> VBlankGo = false;
+boost::lockfree::queue<WriteThing> WriteQueue(128);
```
Comment (MerryMage) [inline src/core/hw/gpu.cpp]: Performance note: Prefer `boost::lockfree::spsc_queue` in this case.

File: src/core/hw/gpu.cpp [hunk 3]
```diff
@@ -565,6 +578,17 @@ void Init() {
     LOG_DEBUG(HW_GPU, "initialized OK");
 }
 
+void Update() {
+    WriteThing thing;
+    while (WriteQueue.pop(thing)) {
```
Comment (MerryMage) [inline src/core/hw/gpu.cpp]: Performance note: Prefer `spsc_queue::consume_all`.

File: src/core/hw/gpu.cpp [hunk 4]
```diff
@@ -565,6 +578,17 @@ void Init() {
...
+        WriteE<u32>(thing.addr, thing.data);
+    }
+    if (VBlankGo) {
```
Comment (MerryMage) [inline src/core/hw/gpu.cpp]: Race condition: It is possible for there to be an elements in the write queue that should have been executed after the VBlank.

Prefer to use the queue for all communication: i.e.: place a sentinel value on the queue.

File: src/core/core.cpp [hunk 5]
```diff
@@ -47,6 +53,15 @@ System::ResultStatus System::RunLoop(int tight_loop) {
         }
     }
 
+    TightLoop += tight_loop;
```
Comment (MerryMage) [inline src/core/core.cpp]: Note: It may be possible for `TightLoop` to grow unboundedly.

**Comment by Subv** [comment]
Signaling the GPU interrupts from the GPU thread can cause race conditions, we have an HLE mutex for cases like this.

**Comment by B3n30** [comment]
Closed, since it is only POC

**Pull request closed without merging**